# Problema de clasificacion
En este ejercicio se pretende clasificar las aplicaciones por descargas. Esto se hara con base a ciertos parametros, que son los siguientes: cantidad de criticas(Reviews) y puntuacion promedio de esas criticas (Rating). Una vez se realice la regresion lineal, se pretende conocer la cantidad de instalaciones que una aplicacion alcanzo, conociendo solamente esos dos datos de parametro. 


In [1]:
import os 
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LinearRegression

In [2]:
data_set = pd.read_csv('googleplaystore-transformed.csv')

In [3]:
data_set.head()

App        Category  Rating  Reviews    Size  Installs  Type  \
0  ¡Ay Caramba!          FAMILY     0.0        0  549000         1  Paid   
1    ¡Ay Metro!            GAME     3.8      489     36M     10000  Free   
2    ¿Es Vegan?  FOOD_AND_DRINK     4.6      438    1.7M     10000  Free   
3        /u/app   COMMUNICATION     4.7      573     53M     10000  Free   
4            .R           TOOLS     4.5      259    203k     10000  Free   

   Price Content Rating         Genres          Last Updated Current Ver  \
0   1.99       Everyone      Education  2014-06-13T00:00:00Z         1.2   
1   0.00   Everyone 10+         Arcade  2015-03-17T00:00:00Z     1.0.3.1   
2   0.00       Everyone   Food & Drink  2017-08-01T00:00:00Z       2.2.3   
3   0.00     Mature 17+  Communication  2018-07-03T00:00:00Z       4.2.4   
4   0.00       Everyone          Tools  2014-09-16T00:00:00Z      1.1.06   

  Android Ver  
0  3.0 and up  
1  4.0 and up  
2  3.0 and up  
3  4.1 and up  
4  1.5 and up

In [4]:
#Cantidad de filas que contiene el dataset
data_set.shape

(9602, 13)

Como se puede observar el dataset tiene un tamaño relativamente grande. 

A continuacion vamos a proceder a realizar nuestros vectores x y y. Los cuales seran nuestros puntos de comparacion en la clasificacion. 

In [5]:
x = data_set[['Rating', 'Reviews']]
y = data_set['Installs']

In [6]:

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, train_size = 0.8, random_state=1)
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(7681, 2)
(7681,)
(1921, 2)
(1921,)


In [7]:
model = LinearRegression()

# fit del modelo 
model.fit(X_train, y_train)

LinearRegression()

In [10]:
# pair the feature names with the coefficients
print(model.intercept_)
list(zip(['Rating', 'Reviews'], model.coef_))

171107.4306207681


[('Rating', 986107.3783209667), ('Reviews', 18.78339400860837)]

Esto significa que el coeficiente de las puntuaciones en las criticas(Rating) equivale a 986107.3783209667 y  cantidad de criticas obtenida es 18.78339400860837. Por lo que podemos interpretar lo siguiente

* Por cada critica que un usuario escribe, hay 18 instalaciones.   

* Por cada *estrella* en la puntuacion equivale a 986107 instalaciones.
Estrella es analogo a la escala del 1 al 5 que existen en la seccion de criticas de la playstore.

Para obtener y se usa la siguiente formula
$$y = 171107.4306207681 +986107.3783209667 \cdot \text{Rating} + 18.78339400860837 \cdot \text{Reviews} $$

[      1   10000   10000 ... 1000000   10000 1000000]
[4474082.34728715 4856062.05845367 3956993.70336154 ...  233413.12136499
 3977303.3675766  5060013.47486204]


In [22]:
from sklearn import metrics
y_pred = model.predict(X_test)

y_set = data_set['Installs']
y_true = y_set.to_numpy()


print('MAE:', metrics.mean_absolute_error(y_test, y_pred))

MAE: 7597367.349987324


In [23]:
print('MSE:', metrics.mean_squared_error(y_test, y_pred))

MSE: 1616618881280933.5


In [24]:
import numpy as np
print('RMSE:',np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

RMSE: 40207199.371268496


In [26]:
rating = input('Ingrese una puntuacion')
review = input('Ingrese la cantidad de criticas')

model = RandomForestClassifier()
model.fit(x, y)
# make a single prediction
row = [[rating,review]]
yhat = model.predict(row)
print('Cantidad de instalaciones predecido: %d' % yhat[0])

Cantidad de instalaciones predecido: 10000000
